sbin/routed/routed.8
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" %sccs.include.redist.man%
.\"
.\"     @(#)routed.8	8.2 (Berkeley) 12/11/93
.\"
.Dd 
.Dt ROUTED 8
.Os BSD 4.2
.Sh NAME
.Nm routed
.Nd network routing daemon
.Sh SYNOPSIS
.Nm routed
.Op Fl d
.Op Fl g
.Op Fl q
.Op Fl s
.Op Fl t
.Op Ar logfile
.Sh DESCRIPTION
.Nm Routed
is invoked at boot time to manage the network routing tables.
The routing daemon uses a variant of the Xerox NS Routing
Information Protocol in maintaining up to date kernel routing
table entries.
It used a generalized protocol capable of use with multiple
address types, but is currently used only for Internet routing
within a cluster of networks.
.Pp
In normal operation
.Nm routed
listens on the
.Xr udp 4
socket for the
.Xr route 8
service (see
.Xr services 5 )
for routing information packets.  If the host is an
internetwork router, it periodically supplies copies
of its routing tables to any directly connected hosts
and networks.
.Pp
When
.Nm routed
is started, it uses the
.Dv SIOCGIFCONF
.Xr ioctl 2
to find those
directly connected interfaces configured into the
system and marked ``up'' (the software loopback interface
is ignored).  If multiple interfaces
are present, it is assumed that the host will forward packets
between networks.
.Nm Routed
then transmits a 
.Em request
packet on each interface (using a broadcast packet if
the interface supports it) and enters a loop, listening
for
.Em request
and
.Em response
packets from other hosts.
.Pp
When a
.Em request
packet is received, 
.Nm routed
formulates a reply based on the information maintained in its
internal tables.  The
.Em response
packet generated contains a list of known routes, each marked
with a ``hop count'' metric (a count of 16, or greater, is
considered ``infinite'').  The metric associated with each
route returned provides a metric
.Em relative to the sender .
.Pp
.Em Response
packets received by
.Nm routed
are used to update the routing tables if one of the following
conditions is satisfied:
.Bl -enum
.It
No routing table entry exists for the destination network
or host, and the metric indicates the destination is ``reachable''
(i.e. the hop count is not infinite).
.It
The source host of the packet is the same as the router in the
existing routing table entry.  That is, updated information is
being received from the very internetwork router through which
packets for the destination are being routed.
.It
The existing entry in the routing table has not been updated for
some time (defined to be 90 seconds) and the route is at least
as cost effective as the current route.
.It
The new route describes a shorter route to the destination than
the one currently stored in the routing tables; the metric of
the new route is compared against the one stored in the table
to decide this.
.El
.Pp
When an update is applied,
.Nm routed
records the change in its internal tables and updates the kernel
routing table.
The change is reflected in the next
.Em response
packet sent.
.Pp
In addition to processing incoming packets,
.Nm routed
also periodically checks the routing table entries.
If an entry has not been updated for 3 minutes, the entry's metric
is set to infinity and marked for deletion.  Deletions are delayed
an additional 60 seconds to insure the invalidation is propagated
throughout the local internet.
.Pp
Hosts acting as internetwork routers gratuitously supply their
routing tables every 30 seconds to all directly connected hosts
and networks.
The response is sent to the broadcast address on nets capable of that function,
to the destination address on point-to-point links, and to the router's
own address on other networks.
The normal routing tables are bypassed when sending gratuitous responses.
The reception of responses on each network is used to determine that the
network and interface are functioning correctly.
If no response is received on an interface, another route may be chosen
to route around the interface, or the route may be dropped if no alternative
is available.
.Pp
Options supported by
.Nm routed :
.Bl -tag -width Ds
.It Fl d
Enable additional debugging information to be logged,
such as bad packets received.
.It Fl g
This flag is used on internetwork routers to offer a route
to the ``default'' destination.
This is typically used on a gateway to the Internet,
or on a gateway that uses another routing protocol whose routes
are not reported to other local routers.
.It Fl s
Supplying this
option forces 
.Nm routed
to supply routing information whether it is acting as an internetwork
router or not.
This is the default if multiple network interfaces are present,
or if a point-to-point link is in use.
.It Fl q
This
is the opposite of the
.Fl s
option.
.It Fl t
If the
.Fl t
option is specified, all packets sent or received are
printed on the standard output.  In addition,
.Nm routed
will not divorce itself from the controlling terminal
so that interrupts from the keyboard will kill the process.
.El
.Pp
Any other argument supplied is interpreted as the name
of file in which 
.Nm routed Ns \'s
actions should be logged.  This log contains information
about any changes to the routing tables and, if not tracing all packets,
a history of recent messages sent and received which are related to
the changed route.
.Pp
In addition to the facilities described above, 
.Nm routed
supports the notion of ``distant''
.Em passive
and 
.Em active
gateways.  When 
.Nm routed
is started up, it reads the file
.Pa /etc/gateways
to find gateways which may not be located using
only information from the
.Dv SIOGIFCONF
.Xr ioctl 2 .
Gateways specified in this manner should be marked passive
if they are not expected to exchange routing information,
while gateways marked active
should be willing to exchange routing information (i.e.
they should have a
.Nm routed
process running on the machine).
Routes through passive gateways are installed in the
kernel's routing tables once upon startup.
Such routes are not included in
any routing information transmitted.
Active gateways are treated equally to network
interfaces.  Routing information is distributed
to the gateway and if no routing information is
received for a period of time, the associated
route is deleted.
Gateways marked
.Em external
are also passive, but are not placed in the kernel
routing table nor are they included in routing updates.
The function of external entries is to inform
.Nm routed
that another routing process
will install such a route, and that alternate routes to that destination
should not be installed.
Such entries are only required when both routers may learn of routes
to the same destination.
.Pp
The 
.Pa /etc/gateways
is comprised of a series of lines, each in
the following format:
.Bd -ragged
.Pf < Cm net No \&|
.Cm host Ns >
.Ar name1
.Cm gateway
.Ar name2
.Cm metric
.Ar value
.Pf < Cm passive No \&|
.Cm active No \&|
.Cm external Ns >
.Ed
.Pp
The 
.Cm net
or
.Cm host
keyword indicates if the route is to a network or specific host.
.Pp
.Ar Name1
is the name of the destination network or host.  This may be a
symbolic name located in
.Pa /etc/networks
or
.Pa /etc/hosts
(or, if started after
.Xr named 8 ,
known to the name server), 
or an Internet address specified in ``dot'' notation; see
.Xr inet 3 .
.Pp
.Ar Name2
is the name or address of the gateway to which messages should
be forwarded.
.Pp
.Ar Value
is a metric indicating the hop count to the destination host
or network.
.Pp
One of the keywords
.Cm passive ,
.Cm active
or
.Cm external
indicates if the gateway should be treated as 
.Em passive
or
.Em active
(as described above),
or whether the gateway is
.Em external
to the scope of the
.Nm routed
protocol.
.Pp
Internetwork routers that are directly attached to the Arpanet or Milnet
should use the Exterior Gateway Protocol
.Pq Tn EGP
to gather routing information
rather then using a static routing table of passive gateways.
.Tn EGP
is required in order to provide routes for local networks to the rest
of the Internet system.
.Sh FILES
.Bl -tag -width /etc/gateways -compact
.It Pa /etc/gateways
for distant gateways
.El
.Sh SEE ALSO
.Xr udp 4 ,
.Xr icmp 4 ,
.Xr XNSrouted 8 ,
.Xr htable 8
.Rs
.%T Internet Transport Protocols
.%R XSIS 028112
.%Q Xerox System Integration Standard
.Re
.Sh BUGS
The kernel's routing tables may not correspond to those of
.Nm routed
when redirects change or add routes.
.Nm Routed
should note any redirects received by reading
the
.Tn ICMP
packets received via a raw socket.
.Pp
.Nm Routed
should incorporate other routing protocols,
such as Xerox
.Tn \&NS
.Pq Xr XNSrouted 8
and
.Tn EGP .
Using separate processes for each requires configuration options
to avoid redundant or competing routes.
.Pp
.Nm Routed
should listen to intelligent interfaces, such as an
.Tn IMP ,
to gather more information.
It does not always detect unidirectional failures in network interfaces
(e.g., when the output side fails).
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.2 .


sbin/routed/routed.c
#ifndef lint
static char sccsid[] = "@(#)routed.c	4.3 05/24/82";
#endif

#include <sys/param.h>
#include <sys/protosw.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <net/in.h>
#define	KERNEL
#include <net/route.h>
#include <net/if.h>
#include <errno.h>
#include <stdio.h>
#include <nlist.h>
#include <signal.h>
#include "rip.h"
#include "router.h"

#define	LOOPBACKNET	0177
/* casts to keep lint happy */
#define	insque(q,p)	_insque((caddr_t)q,(caddr_t)p)
#define	remque(q)	_remque((caddr_t)q)
#define equal(a1, a2) \
	(bcmp((caddr_t)(a1), (caddr_t)(a2), sizeof (struct sockaddr)) == 0)
#define	min(a,b)	((a)>(b)?(b):(a))

struct nlist nl[] = {
#define	N_IFNET		0
	{ "_ifnet" },
	0,
};

struct	sockaddr_in myaddr = { AF_INET, IPPORT_ROUTESERVER };

int	s;
int	kmem = -1;
int	supplier;		/* process should supply updates */
int	initializing;		/* stem off broadcast() calls */
int	install = 0;		/* if 1 call kernel */
int	lookforinterfaces = 1;
int	performnlist = 1;
int	timeval;
int	timer();
int	cleanup();
int	trace = 0;

char	packet[MAXPACKETSIZE];

extern char *malloc();
extern int errno, exit();

main(argc, argv)
	int argc;
	char *argv[];
{
	int cc;
	struct sockaddr from;
	
#ifdef notdef
	{   int t = open("/dev/tty", 2);
	    if (t >= 0) {
		ioctl(t, TIOCNOTTY, 0);
		close(t);
	    }
	}
#endif
	if (trace) {
		(void) freopen("/etc/routerlog", "a", stdout);
		(void) dup2(fileno(stdout), 2);
		setbuf(stdout, NULL);
	}
#ifdef vax
	myaddr.sin_port = htons(myaddr.sin_port);
#endif
again:
	s = socket(SOCK_DGRAM, 0, &myaddr, 0);
	if (s < 0) {
		perror("socket");
		sleep(30);
		goto again;
	}
	rtinit();
	getothers();
	initializing = 1;
	getinterfaces();
	initializing = 0;
	request();

	argv++, argc--;
	while (argc > 0) {
		if (strcmp(*argv, "-s") == 0)
			supplier = 1;
		else if (strcmp(*argv, "-q") == 0)
			supplier = 0;
		argv++, argc--;
	}
	sigset(SIGALRM, timer);
	timer();

	/*
	 * Listen for routing packets
	 */
	for (;;) {
		cc = receive(s, &from, packet, sizeof (packet));
		if (cc <= 0) {
			if (cc < 0 && errno != EINTR)
				perror("receive");
			continue;
		}
		sighold(SIGALRM);
		rip_input(&from, cc);
		sigrelse(SIGALRM);
	}
}

/*
 * Look in a file for any gateways we should configure
 * outside the directly connected ones.  This is a kludge,
 * but until we can find out about gateways on the "other side"
 * of the ARPANET using GGP, it's a must.
 *
 * We don't really know the distance to the gateway, so we
 * assume it's a neighbor.
 */
getothers()
{
	struct sockaddr_in dst, gate;
	FILE *fp = fopen("/etc/gateways", "r");
	struct rt_entry *rt;

	if (fp == NULL)
		return;
	bzero((char *)&dst, sizeof (dst));
	bzero((char *)&gate, sizeof (gate));
	dst.sin_family = AF_INET;
	gate.sin_family = AF_INET;
	while (fscanf(fp, "%x %x", &dst.sin_addr.s_addr, 
	   &gate.sin_addr.s_addr) != EOF) {
		rtadd((struct sockaddr *)&dst, (struct sockaddr *)&gate, 1);
		rt = rtlookup((struct sockaddr *)&dst);
		if (rt)
			rt->rt_flags |= RTF_SILENT;
	}
	fclose(fp);
}

/*
 * Timer routine:
 *
 * o handle timers on table entries,
 * o invalidate entries which haven't been updated in a while,
 * o delete entries which are too old,
 * o retry ioctl's which weren't successful the first
 *   time due to the kernel entry being busy
 * o if we're an internetwork router, supply routing updates
 *   periodically
 */
timer()
{
	register struct rt_hash *rh;
	register struct rt_entry *rt;
	struct rt_hash *base = hosthash;
	int doinghost = 1;

	if (trace)
		printf(">>> time %d >>>\n", timeval);
again:
	for (rh = base; rh < &base[ROUTEHASHSIZ]; rh++) {
		rt = rh->rt_forw;
		for (; rt != (struct rt_entry *)rh; rt = rt->rt_forw) {

			/*
			 * If the host is indicated to be
			 * "silent" (i.e. it's one we got
			 * from the initialization file),
			 * don't time out it's entry.
			 */
			if (rt->rt_flags & RTF_SILENT)
				continue;
			log("", rt);
			rt->rt_timer += TIMER_RATE;

			/*
			 * If the entry should be deleted
			 * attempt to do so and reclaim space.
			 */
			if (rt->rt_timer >= GARBAGE_TIME ||
			  (rt->rt_flags & RTF_DELRT)) {
				rt = rt->rt_forw;
				rtdelete(rt->rt_back);
				rt = rt->rt_back;
				continue;
			}

			/*
			 * If we haven't heard from the router
			 * in a long time indicate the route is
			 * hard to reach.
			 */
			if (rt->rt_timer >= EXPIRE_TIME)
				rt->rt_metric = HOPCNT_INFINITY;
			if (rt->rt_flags & RTF_CHGRT)
				if (!ioctl(s, SIOCCHGRT,(char *)&rt->rt_hash) ||
				  --rt->rt_retry == 0)
					rt->rt_flags &= ~RTF_CHGRT;

			/*
			 * Try to add the route to the kernel tables.
			 * If this fails because the entry already exists
			 * (perhaps because someone manually added it)
			 * change the add to a change.  If the operation
			 * fails otherwise (likely because the entry is
			 * in use), retry the operation a few more times.
			 */
			if (rt->rt_flags & RTF_ADDRT) {
				if (!ioctl(s, SIOCADDRT,(char *)&rt->rt_hash)) {
					if (errno == EEXIST) {
						rt->rt_flags &= ~RTF_ADDRT;
						rt->rt_flags |= RTF_CHGRT;
						rt->rt_retry =
						    (EXPIRE_TIME/TIMER_RATE);
						continue;
					}
					if (--rt->rt_retry)
						continue;
				}
				rt->rt_flags &= ~RTF_ADDRT;
			}
		}
	}
	if (doinghost) {
		doinghost = 0;
		base = nethash;
		goto again;
	}
	timeval += TIMER_RATE;
	if (lookforinterfaces && (timeval % CHECK_INTERVAL) == 0)
		getinterfaces();
	if (supplier && (timeval % SUPPLY_INTERVAL) == 0)
		supplyall();
	if (trace)
		printf("<<< time %d <<<\n", timeval);
	alarm(TIMER_RATE);
}

/*
 * Find the network interfaces attached to this machine.
 * The info is used to:
 *
 * (1) initialize the routing tables, as done by the kernel.
 * (2) ignore incoming packets we send.
 * (3) figure out broadcast capability and addresses.
 * (4) figure out if we're an internetwork gateway.
 *
 * We don't handle anything but Internet addresses.
 */
getinterfaces()
{
	struct ifnet *ifp;
	struct ifnet ifstruct, *next;
	struct sockaddr_in net;
	register struct sockaddr *dst;
	int nets;

	if (performnlist) {
		nlist("/vmunix", nl);
		if (nl[N_IFNET].n_value == 0) {
			performnlist++;
			printf("ifnet: not in namelist\n");
			return;
		}
		performnlist = 0;
	}
	if (kmem < 0) {
		kmem = open("/dev/kmem", 0);
		if (kmem < 0) {
			perror("/dev/kmem");
			return;
		}
	}
	if (lseek(kmem, (long)nl[N_IFNET].n_value, 0) == -1 ||
	    read(kmem, (char *)&ifp, sizeof (ifp)) != sizeof (ifp)) {
		performnlist = 1;
		return;
	}
	bzero((char *)&net, sizeof (net));
	net.sin_family = AF_INET;
	nets = 0;
	while (ifp) {
		if (lseek(kmem, (long)ifp, 0) == -1 ||
		  read(kmem, (char *)&ifstruct, sizeof (ifstruct)) !=
		  sizeof (ifstruct)) {
			perror("read");
			performnlist = 1;
			break;
		}
		ifp = &ifstruct;
		if ((ifp->if_flags & IFF_UP) == 0) {
			lookforinterfaces++;
	skip:
			ifp = ifp->if_next;
			continue;
		}
		if (ifp->if_addr.sa_family != AF_INET)
			goto skip;
		/* ignore software loopback networks. */
		if (ifp->if_net == LOOPBACKNET)
			goto skip;
		/* check if we already know about this one */
		if (if_ifwithaddr(&ifstruct.if_addr)) {
			nets++;
			goto skip;
		}
		ifp = (struct ifnet *)malloc(sizeof (struct ifnet));
		if (ifp == 0) {
			printf("routed: out of memory\n");
			break;
		}
		bcopy((char *)&ifstruct, (char *)ifp, sizeof (struct ifnet));

		/*
		 * Count the # of directly connected networks
		 * and point to point links which aren't looped
		 * back to ourself.  This is used below to
		 * decide if we should be a routing "supplier".
		 */
		if ((ifp->if_flags & IFF_POINTOPOINT) == 0 ||
		    if_ifwithaddr(&ifp->if_dstaddr) == 0)
			nets++;

		if (ifp->if_flags & IFF_POINTOPOINT)
			dst = &ifp->if_dstaddr;
		else {
			net.sin_addr = if_makeaddr(ifp->if_net, INADDR_ANY);
			dst = (struct sockaddr *)&net;
		}
		next = ifp->if_next;
		ifp->if_next = ifnet;
		ifnet = ifp;
		if (rtlookup(dst) == 0)
			rtadd(dst, &ifp->if_addr, 0);
		ifp = next;
	}
	supplier = nets > 1;
}

/*
 * Send a request message to all directly
 * connected hosts and networks.
 */
request()
{
	register struct rip *msg = (struct rip *)packet;

	msg->rip_cmd = RIPCMD_REQUEST;
	msg->rip_nets[0].rip_dst.sa_family = AF_UNSPEC;
	msg->rip_nets[0].rip_metric = HOPCNT_INFINITY;
	sendall();
}

/*
 * Broadcast a new, or modified, routing table entry
 * to all directly connected hosts and networks.
 */
broadcast(entry)
	struct rt_entry *entry;
{
	struct rip *msg = (struct rip *)packet;

	log("broadcast", entry);
	msg->rip_cmd = RIPCMD_RESPONSE;
	msg->rip_nets[0].rip_dst = entry->rt_dst;
	msg->rip_nets[0].rip_metric = min(entry->rt_metric+1, HOPCNT_INFINITY);
	sendall();
}

sendall()
{
	register struct rt_hash *rh;
	register struct rt_entry *rt;
	register struct sockaddr *dst;
	struct rt_hash *base = hosthash;
	int doinghost = 1;

again:
	for (rh = base; rh < &base[ROUTEHASHSIZ]; rh++)
	for (rt = rh->rt_forw; rt != (struct rt_entry *)rh; rt = rt->rt_forw) {
		if ((rt->rt_flags & RTF_SILENT) || rt->rt_metric > 0)
			continue;
		if (rt->rt_ifp && (rt->rt_ifp->if_flags & IFF_BROADCAST))
			dst = &rt->rt_ifp->if_broadaddr;
		else
			dst = &rt->rt_gateway;
		(*afswitch[dst->sa_family].af_output)(dst, sizeof (struct rip));
	}
	if (doinghost) {
		base = nethash;
		doinghost = 0;
		goto again;
	}
}

/*
 * Supply all directly connected neighbors with the
 * current state of the routing tables.
 */
supplyall()
{
	register struct rt_entry *rt;
	register struct rt_hash *rh;
	register struct sockaddr *dst;
	struct rt_hash *base = hosthash;
	int doinghost = 1;

again:
	for (rh = base; rh < &base[ROUTEHASHSIZ]; rh++)
	for (rt = rh->rt_forw; rt != (struct rt_entry *)rh; rt = rt->rt_forw) {
		if ((rt->rt_flags & RTF_SILENT) || rt->rt_metric > 0)
			continue;
		if (rt->rt_ifp && (rt->rt_ifp->if_flags & IFF_BROADCAST))
			dst = &rt->rt_ifp->if_broadaddr;
		else
			dst = &rt->rt_gateway;
		log("supply", rt);
		supply(dst);
	}
	if (doinghost) {
		base = nethash;
		doinghost = 0;
		goto again;
	}
}

/*
 * Supply routing information to target "sa".
 */
supply(sa)
	struct sockaddr *sa;
{
	struct rip *msg = (struct rip *)packet;
	struct netinfo *n = msg->rip_nets;
	register struct rt_hash *rh;
	register struct rt_entry *rt;
	struct rt_hash *base = hosthash;
	int doinghost = 1, size;
	int (*output)() = afswitch[sa->sa_family].af_output;

	msg->rip_cmd = RIPCMD_RESPONSE;
again:
	for (rh = base; rh < &base[ROUTEHASHSIZ]; rh++)
	for (rt = rh->rt_forw; rt != (struct rt_entry *)rh; rt = rt->rt_forw) {

		/*
		 * Flush packet out if not enough room for
		 * another routing table entry.
		 */
		size = (char *)n - packet;
		if (size > MAXPACKETSIZE - sizeof (struct netinfo)) {
			(*output)(sa, size);
			n = msg->rip_nets;
		}
		n->rip_dst = rt->rt_dst;
		n->rip_metric = min(rt->rt_metric + 1, HOPCNT_INFINITY);
		n++;
	}
	if (doinghost) {
		doinghost = 0;
		base = nethash;
		goto again;
	}
	if (n != msg->rip_nets)
		(*output)(sa, (char *)n - packet);
}

/*
 * Respond to a routing info request.
 */
rip_respond(from, size)
	struct sockaddr *from;
	int size;
{
	register struct rip *msg = (struct rip *)packet;
	struct netinfo *np = msg->rip_nets;
	struct rt_entry *rt;
	int newsize = 0;
	
	size -= 4 * sizeof (char);
	while (size > 0) {
		if (size < sizeof (struct netinfo))
			break;
		size -= sizeof (struct netinfo);
		if (np->rip_dst.sa_family == AF_UNSPEC &&
		    np->rip_metric == HOPCNT_INFINITY && size == 0) {
			supply(from);
			return;
		}
		rt = rtlookup(&np->rip_dst);
		np->rip_metric = rt == 0 ?
			HOPCNT_INFINITY : min(rt->rt_metric+1, HOPCNT_INFINITY);
		np++, newsize += sizeof (struct netinfo);
	}
	if (newsize > 0) {
		msg->rip_cmd = RIPCMD_RESPONSE;
		newsize += sizeof (int);
		(*afswitch[from->sa_family].af_output)(from, newsize);
	}
}

/*
 * Handle an incoming routing packet.
 */
rip_input(from, size)
	struct sockaddr *from;
	int size;
{
	register struct rip *msg = (struct rip *)packet;
	struct rt_entry *rt;
	struct netinfo *n;

	if (msg->rip_cmd != RIPCMD_RESPONSE &&
	    msg->rip_cmd != RIPCMD_REQUEST)
		return;

	if (msg->rip_cmd == RIPCMD_RESPONSE &&
	    (*afswitch[from->sa_family].af_portmatch)(from) == 0)
		return;
	if (msg->rip_cmd == RIPCMD_REQUEST) {
		rip_respond(from, size);
		return;
	}

	/*
	 * Process updates.
	 * Extraneous information like Internet ports
	 * must first be purged from the sender's address for
	 * pattern matching below.
	 */
	(*afswitch[from->sa_family].af_canon)(from);
	if (trace)
		printf("input from %x\n", from->sin_addr);
	/*
	 * If response packet is from ourselves, use it only
	 * to reset timer on entry.  Otherwise, we'd believe
	 * it as gospel (since it comes from the router) and
	 * unknowingly update the metric to show the outgoing
	 * cost (higher than our real cost).  I guess the protocol
	 * spec doesn't address this because Xerox Ethernets
	 * don't hear their own broadcasts?
	 */
	if (if_ifwithaddr(from)) {
		rt = rtlookup(from);
		if (rt)
			rt->rt_timer = 0;
		return;
	}
	size -= 4 * sizeof (char);
	n = msg->rip_nets;
	for (; size > 0; size -= sizeof (struct netinfo), n++) {
		if (size < sizeof (struct netinfo))
			break;
		if (trace)
			printf("dst %x hc %d...", n->rip_dst.sin_addr,
				n->rip_metric);
		rt = rtlookup(&n->rip_dst);

		/*
		 * Unknown entry, add it to the tables only if
		 * its interesting.
		 */
		if (rt == 0) {
			if (n->rip_metric < HOPCNT_INFINITY)
				rtadd(&n->rip_dst, from, n->rip_metric);
			if (trace)
				printf("new\n");
			continue;
		}

		if (trace)
			printf("ours: gate %x hc %d timer %d\n",
			rt->rt_gateway.sin_addr,
			rt->rt_metric, rt->rt_timer);
		/*
		 * Update the entry if one of the following is true:
		 *
		 * (1) The update came directly from the gateway.
		 * (2) A shorter path is provided.
		 * (3) The entry hasn't been updated in a while
		 *     and a path of equivalent cost is offered
		 *     (with the cost finite).
		 */
		if (equal(from, &rt->rt_gateway) ||
		    rt->rt_metric > n->rip_metric ||
		    (rt->rt_timer > (EXPIRE_TIME/2) &&
		    rt->rt_metric == n->rip_metric &&
		    rt->rt_metric < HOPCNT_INFINITY)) {
			rtchange(rt, from, n->rip_metric);
			rt->rt_timer = 0;
		}
	}
}

/*
 * Lookup an entry to the appropriate dstination.
 */
struct rt_entry *
rtlookup(dst)
	struct sockaddr *dst;
{
	register struct rt_entry *rt;
	register struct rt_hash *rh;
	register int hash, (*match)();
	struct afhash h;
	int af = dst->sa_family, doinghost = 1;

	if (af >= AF_MAX)
		return (0);
	(*afswitch[af].af_hash)(dst, &h);
	hash = h.afh_hosthash;
	rh = &hosthash[hash % ROUTEHASHSIZ];

again:
	for (rt = rh->rt_forw; rt != (struct rt_entry *)rh; rt = rt->rt_forw) {
		if (rt->rt_hash != hash)
			continue;
		if (doinghost) {
			if (equal(&rt->rt_dst, dst))
				return (rt);
		} else {
			if (rt->rt_dst.sa_family == af &&
			    (*match)(&rt->rt_dst, dst))
				return (rt);
		}
	}
	if (doinghost) {
		doinghost = 0;
		hash = h.afh_nethash;
		match = afswitch[af].af_netmatch;
		rh = &nethash[hash % ROUTEHASHSIZ];
		goto again;
	}
	return (0);
}

rtinit()
{
	register struct rt_hash *rh;

	for (rh = nethash; rh < &nethash[ROUTEHASHSIZ]; rh++)
		rh->rt_forw = rh->rt_back = (struct rt_entry *)rh;
	for (rh = hosthash; rh < &hosthash[ROUTEHASHSIZ]; rh++)
		rh->rt_forw = rh->rt_back = (struct rt_entry *)rh;
}

/*
 * Add a new entry.
 */
rtadd(dst, gate, metric)
	struct sockaddr *dst, *gate;
	short metric;
{
	struct afhash h;
	register struct rt_entry *rt;
	struct rt_hash *rh;
	int af = dst->sa_family, flags, hash;

	if (af >= AF_MAX)
		return;
	(*afswitch[af].af_hash)(dst, &h);
	flags = (*afswitch[af].af_checkhost)(dst) ? RTF_HOST : 0;
	if (flags & RTF_HOST) {
		hash = h.afh_hosthash;
		rh = &hosthash[hash % ROUTEHASHSIZ];
	} else {
		hash = h.afh_nethash;
		rh = &nethash[hash % ROUTEHASHSIZ];
	}
	rt = (struct rt_entry *)malloc(sizeof (*rt));
	if (rt == 0)
		return;
	rt->rt_hash = hash;
	rt->rt_dst = *dst;
	rt->rt_gateway = *gate;
	rt->rt_metric = metric;
	rt->rt_timer = 0;
	rt->rt_flags = RTF_UP | flags;
	rt->rt_ifp = if_ifwithnet(&rt->rt_gateway);
	if (metric == 0)
		rt->rt_flags |= RTF_DIRECT;
	insque(rt, rh);
	log("add", rt);
	if (initializing)
		return;
	if (supplier)
		broadcast(rt);
	if (install) {
		rt->rt_flags |= RTF_ADDRT;
		rt->rt_retry = EXPIRE_TIME/TIMER_RATE;
	}
}

/*
 * Look to see if a change to an existing entry
 * is warranted; if so, make it.
 */
rtchange(rt, gate, metric)
	struct rt_entry *rt;
	struct sockaddr *gate;
	short metric;
{
	int change = 0;

	if (!equal(&rt->rt_gateway, gate)) {
		rt->rt_gateway = *gate;
		change++;
	}

	/*
	 * If the hop count has changed, adjust
	 * the flags in the routing table entry accordingly.
	 */
	if (metric != rt->rt_metric) {
		if (rt->rt_metric == 0)
			rt->rt_flags &= ~RTF_DIRECT;
		rt->rt_metric = metric;
		if (metric >= HOPCNT_INFINITY)
			rt->rt_flags &= ~RTF_UP;
		else
			rt->rt_flags |= RTF_UP;
		change++;
	}

	if (!change)
		return;
	if (supplier)
		broadcast(rt);
	log("change", rt);
	if (install) {
		rt->rt_flags |= RTF_CHGRT;
		rt->rt_retry = EXPIRE_TIME/TIMER_RATE;
	}
}

/*
 * Delete a routing table entry.
 */
rtdelete(rt)
	struct rt_entry *rt;
{
	log("delete", rt);
	if (install)
		if (ioctl(s, SIOCDELRT, (char *)&rt->rt_hash) &&
		  errno == EBUSY)
			rt->rt_flags |= RTF_DELRT;
	remque(rt);
	free((char *)rt);
}

log(operation, rt)
	char *operation;
	struct rt_entry *rt;
{
	time_t t = time(0);
	struct sockaddr_in *dst, *gate;
	static struct flagbits {
		int	t_bits;
		char	*t_name;
	} bits[] = {
		{ RTF_UP,	"UP" },
		{ RTF_DIRECT,	"DIRECT" },
		{ RTF_HOST,	"HOST" },
		{ RTF_DELRT,	"DELETE" },
		{ RTF_CHGRT,	"CHANGE" },
		{ RTF_SILENT,	"SILENT" },
		{ 0 }
	};
	register struct flagbits *p;
	register int first;
	char *cp;

	if (trace == 0)
		return;
	printf("%s ", operation);
	dst = (struct sockaddr_in *)&rt->rt_dst;
	gate = (struct sockaddr_in *)&rt->rt_gateway;
	printf("dst %x, router %x, metric %d, flags ",
		dst->sin_addr, gate->sin_addr, rt->rt_metric);
	cp = "%s";
	for (first = 1, p = bits; p->t_bits > 0; p++) {
		if ((rt->rt_flags & p->t_bits) == 0)
			continue;
		printf(cp, p->t_name);
		if (first) {
			cp = "|%s";
			first = 0;
		}
	}
	putchar('\n');
}

struct ifnet *
if_ifwithaddr(addr)
	struct sockaddr *addr;
{
	register struct ifnet *ifp;

#define	same(a1, a2) \
	(bcmp((caddr_t)((a1)->sa_data), (caddr_t)((a2)->sa_data), 14) == 0)
	for (ifp = ifnet; ifp; ifp = ifp->if_next) {
		if (ifp->if_addr.sa_family != addr->sa_family)
			continue;
		if (same(&ifp->if_addr, addr))
			break;
		if ((ifp->if_flags & IFF_BROADCAST) &&
		    same(&ifp->if_broadaddr, addr))
			break;
	}
	return (ifp);
#undef same
}

struct ifnet *
if_ifwithnet(addr)
	register struct sockaddr *addr;
{
	register struct ifnet *ifp;
	register int af = addr->sa_family;
	register int (*netmatch)();

	if (af >= AF_MAX)
		return (0);
	netmatch = afswitch[af].af_netmatch;
	for (ifp = ifnet; ifp; ifp = ifp->if_next) {
		if (af != ifp->if_addr.sa_family)
			continue;
		if ((*netmatch)(addr, &ifp->if_addr))
			break;
	}
	return (ifp);
}

struct in_addr
if_makeaddr(net, host)
	int net, host;
{
	u_long addr;

	if (net < 128)
		addr = (net << 24) | host;
	else if (net < 65536)
		addr = (net << 16) | host;
	else
		addr = (net << 8) | host;
#ifdef vax
	addr = htonl(addr);
#endif
	return (*(struct in_addr *)&addr);
}
